research.rst

```diff
@@ -0,0 +1,315 @@
+of Ansible Vault, a symmetric encryption scheme exposed through the
+ansible command-line client.  But this just begs the question of how
+to share the key to decrypt the Vault files amongst the developers who
+need it.  We could share this key via "sneakernet" or on the private
+intranet, but a more convenient and secure method may be to use a
+shared password in LastPass.  The Ansible Vault password file can be
+executable, so we could write code using of one of the LastPass API
+libraries to fetch the key.
+
+So, how should we make use of Ansible Vault-encrypted files?
+It isn't possible to use Vault only on lines or sections, it has to be
+whole files.  So it's recommended to split out only those variables
+that need to be secret into their own files, to minimize the opaque
+binary blobs that get checked into the repo.  The structure would look
+something like this,
+
+::
+
+    group_vars/
+        prod/
+            secrets.yml
+            non-secrets.yml
+        staging/
+            secrets.yml
+            non-secrets.yml
+        dev/
+            secrets.yml
+            non-secrets.yml
+
+
+and then the ``non-secrets.yml`` files make use of variables defined
+in ``secrets.yml``:
+
+.. code-block:: yaml
+
+    postgres_host: localhost
+    postgres_database: pg_database
+    postgres_user: pg_user
+    postgres_password: {{ VAULT_POSTGRES_PASSWORD }}
+    postgres_port: 5432
+
+
+Also, in the playbook and role tasks, make sure to set ``no_log:
+true`` so that the secrets don't get echoed to the console when the
+verbosity is turned up.
+
+.. code-block:: yaml
+
+    - name: VPN Server | Load VPN secret keys
+      include_vars: "vpn-secrets.yml"
+      no_log: true
+
+
+For ease of use, we could do away with the secret/non-secret file
+split for the dev environment.
+
+Other possible options for sharing the Vault key:
+
+- keep on the intranet (either vault files or keys for files)
+- HashiCorp's Vault
+- consul
+- KeePassX
+
+Sources:
+
+- http://www.slideshare.net/excellaco/using-ansible-vault-to-protect-your-secrets
+
+
+Configuration and Customization
+-------------------------------
+
+Ansible will look relative to the playbook directory or the inventory
+directory for variable files and other such files, as well as in the
+appropriate directories inside roles.  So in order to configure a
+project, it is sufficient to have a set of directories named according
+to convention that will contain needed configuration variables.  A
+likely possibility is
+
+::
+
+    django-project-template/
+        inventory/
+            group_vars/
+            host_vars/
+
+
+This has mostly been done already in the ``use_tequila`` branch,
+though some adjustments should be made in order to follow the
+recommended secrets-vs-non-secrets structure.
+
+The ``ansible.cfg`` that ships with the project will need to define
+the inventory location.
+
+Since the relevant playbook(s) for a project will ship inside that
```

Comment:
This isn't how the current tequila deploy script runs deploys - it uses a static playbook file included in tequila. I'm torn between saving every project developer from having to maintain valid playbooks, and providing the flexibility of changing the playbook, but I think if we can keep a static playbook and use variables for all customization, that'll be easier to use in the long run than splitting up the configuration between variables and custom playbooks.